### ISSUE TITLE
Ambiguous methods in API

### ISSUE THREAD

**Issue by danieldietrich** [vavr-io/vavr#1916]
We need to keep API.* as simple as possible - less is more.

For example `SortedSet(Comparator, T...)` and `SortedSet(T...)` are ambiguous because T is inferred to be the upper bound to match all parameters (which might be Object).

<img width="475" alt="ambiguous-api" src="https://cloud.githubusercontent.com/assets/743833/24080385/7b87a97e-0c9e-11e7-8f10-19d4a5964d39.png">

I used JDK 1.8.0_121

**Comment by danieldietrich** [comment]
Interestingly this is also ambiguous:

```java
Set<Integer> set2 = TreeSet.of((i, j) -> j - i, 1, 2, 3);
```

Is this a problem that needs to be addressed in 3.0.0? E.g. is our API generally broken when optional lambdas are passed as first parameter to static factory methods?

Did the compiler behaviour change in recent JDK versions?

**Comment by danieldietrich** [comment]
This is possible/compiles fine (without type-hint):

```java
Comparator<Integer> intcomp = (i, j) -> j - i;
Set<Integer> set1 = TreeSet.of(intcomp, 1, 2, 3);
Set<Integer> set2 = TreeSet.of(Comparator.reverseOrder(), 1, 2, 3);
```

It is not a bug, it is how Java's type system behaves.

#WontFixIt
